### DISCUSSION TITLE
Why doesn't `no-mixed-operators` support the logical not operator?

### DISCUSSION THREAD

**Discussion by yossizahn** [eslint/eslint#16447] [Q&A]
Hi,
I was wondering why the `no-mixed-operators` rule doesn't support the logical not (`!`) operator?
This would enable catching errors like this:
```
if (!a === b) // oops... should have been `a !== b` or `!(a === b)`
```
I am wondering also if adding support for`!` would then allow `no-mixed-operators` to subsume the functionality of `no-unsafe-negation`

TIA

**Comment by mdjermanovic** [comment]
Hi @yossizahn!

> I was wondering why the `no-mixed-operators` rule doesn't support the logical not (`!`) operator?

As stated in the [docs](https://eslint.org/docs/latest/rules/no-mixed-operators), this rule only targets binary operators.

> I am wondering also if adding support for not would then allow `no-mixed-operators` to subsume the functionality of `no-unsafe-negation`

I think it's better to have these two as separate rules because `no-unsafe-negation` catches possible errors (and is in `eslint:recommended`), whereas `no-mixed-operators` is a stylistic rule.

**Reply by yossizahn** [reply]
Thanks @mdjermanovic!

> As stated in the [docs](https://eslint.org/docs/latest/rules/no-mixed-operators), this rule only targets binary operators.

Is there any reason why the rule is more specifically applicable to binary operators? Would it make sense to broaden the rule to target unary operators too?

> I think it's better to have these two as separate rules because no-unsafe-negation catches possible errors (and is in eslint:recommended), whereas no-mixed-operators is a stylistic rule.

I see. 
I am still somewhat unconvinced. They seem very close in spirit. The source of the error in `no-unsafe-negation` is the fact that the code ignores the rules of operator precedence. IIUC, this is also the point of `no-mixed-operators` i.e. to clarify the precedence.

There is already [a stale issue](https://github.com/eslint/eslint/issues/14802) about adding the `===` operator to the `no-unsafe-negation` rule. For some reason it didn't take off.

**Reply by yossizahn** [reply]
> As stated in the [docs](https://eslint.org/docs/latest/rules/no-mixed-operators), this rule only targets binary operators.

Actually bitwise not `~` is supported 🤷

**Reply by mdjermanovic** [reply]
> Actually bitwise not `~` is supported

Can you test if it's actually supported? I think it's a mistake in the documentation.

**Reply by aryehb** [reply]
It appears it's not actually supported, but there's no validation error.

https://eslint.org/play/#eyJ0ZXh0IjoiY29uc3QgZm9vID0gMSArIDIgKiAzO1xuY29uc3QgYSA9IH4xIHwgMiAmIDM7XG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7Im5vLW1peGVkLW9wZXJhdG9ycyI6WyJlcnJvciIseyJncm91cHMiOltbIisiLCItIiwiKiIsIi8iLCIlIiwiKioiXSxbIiYiLCJ8IiwiXiIsIn4iLCI8PCIsIj4+IiwiPj4+Il0sWyI9PSIsIiE9IiwiPT09IiwiIT09IiwiPiIsIj49IiwiPCIsIjw9Il0sWyImJiIsInx8Il0sWyJpbiIsImluc3RhbmNlb2YiXV0sImFsbG93U2FtZVByZWNlZGVuY2UiOnRydWV9XX0sImVudiI6eyJlczYiOnRydWV9fX0=

**Reply by yossizahn** [reply]
Indeed the term `unaryExpression` does not show up in the source code of the rule. It is only looking at binary/logical/conditional expressions

**Reply by mdjermanovic** [reply]
> Is there any reason why the rule is more specifically applicable to binary operators? Would it make sense to broaden the rule to target unary operators too?

More details on the design of this rule can be found in the original issue https://github.com/eslint/eslint/issues/6023. It might make sense to include unary operators, but given our policy about [stylistic rule updates](https://github.com/eslint/eslint#stylistic-rule-updates), we're not adding new options to the core `no-mixed-operators` rule. You can always make a custom rule based on the core rule and add new features to it.

> The source of the error in `no-unsafe-negation` is the fact that the code ignores the rules of operator precedence. IIUC, this is also the point of `no-mixed-operators` i.e. to clarify the precedence

They are similar in that they relate to operator precedence, but code reported by `no-unsafe-negation` is considered a possible error, as it is highly unlikely that the intent was to check `<boolean> in something` or `<boolean> < something`. Code reported by `no-mixed-operators`, such as `a + b * c`, is not considered a possible error. Requiring parentheses to clarify the precedence in this case is a stylistic preference. Separate rules allow for catching possible errors while not enforcing stylistic preferences.

> There is already [a stale issue](https://github.com/eslint/eslint/issues/14802) about adding the `===` operator to the `no-unsafe-negation` rule. For some reason it didn't take off.

The reason is that `<boolean> === something` is not considered a possible error, and therefore checking `!foo === bar` is not in line with the purpose of this rule, which is to catch possible errors.
